File: src/CSharp/Main/BurnTest/GarbageHolder.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;

namespace GCBurn.BurnTest
{
    // Non thread safe type!
    public class GarbageHolder
    {
        public static readonly long TicksPerReleaseCycle = Stopwatch.Frequency / 1_000_000; // 1M / second
        public const int BucketCount = 10;
        public const int BucketSize = 10;

        private readonly List<object>[][] _buckets;
        private readonly long[] _collected;
        private long _startTimestamp;

        public GarbageHolder()
        {
            _buckets = new List<object>[BucketCount][];
            _collected = new long[BucketCount];
            Reset();
        }

        public void Reset()
        {
            for (var i = 0; i < _buckets.Length; i++)
                _buckets[i] = Enumerable.Range(0, BucketSize).Select(_ => new List<object>()).ToArray();
        }

        public void Start()
        {
            _startTimestamp = Stopwatch.GetTimestamp();
            for (var i = 0; i < _collected.Length; i++)
                _collected[i] = 0;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddGarbage(object garbage, sbyte bucketIndex, sbyte generationIndex)
        {
            if (bucketIndex == 0 && generationIndex == 0)
                // Zero hold time = don't hold it
                return;
            _buckets[bucketIndex][generationIndex].Add(garbage);
        }


        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void TryReleaseGarbage()
        {
            var collectionIndex = (Stopwatch.GetTimestamp() - _startTimestamp) / TicksPerReleaseCycle;
            TryReleaseGarbage(0, collectionIndex);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void TryReleaseGarbage(int bucketIndex, long generationCount)
        {
            while (generationCount != 0 && bucketIndex < _buckets.Length) {
                ref var collected = ref _collected[bucketIndex]; // ref is just to speed this up
                var remaining = generationCount - collected;
                collected = generationCount;
                if (remaining >= BucketSize)
                    remaining = BucketSize;
                var bucket = _buckets[bucketIndex];
                Array.Copy(bucket, remaining, bucket, 0, BucketSize - remaining);
                for (var i = remaining; i < BucketSize; i++)
                    bucket[i] = new List<object>();

                // Kind of a tail call
                bucketIndex++;
                generationCount /= BucketSize;
            }
        }
    }
}

File: src/Go/burn_test/garbage_holder.go
package burn_test

import (
	. "../common"
	. "../common/nanotime"
)

const NanosecondsPerReleaseCycle = Giga / 1e6 // 1M / second
const BucketCount = 10
const BucketSize = 10

type GarbageHolder struct {
	buckets   [BucketCount][BucketSize][][]int64
	collected []int
	startTime int64
}

func NewGarbageHolder() *GarbageHolder {
	g := &GarbageHolder{}
	g.collected = make([]int, BucketCount, BucketCount)
	return g
}

func (g *GarbageHolder) Reset() {
	for i := range g.buckets {
		for j := range g.buckets[i] {
			g.buckets[i][j] = nil
		}
	}
}

func (g *GarbageHolder) Start() {
	g.startTime = Nanotime().Nanoseconds()
	for i := range g.collected {
		g.collected[i] = 0
	}
}

func (g *GarbageHolder) AddGarbage(garbage []int64, bucket, generation int8) {
	if bucket == 0 && generation == 0 {
		// Zero hold time = don't hold it
		return
	}
	g.buckets[bucket][generation] = append(g.buckets[bucket][generation], garbage)
}

func (g *GarbageHolder) TryReleaseGarbage() {
	collectionIdx := int((Nanotime().Nanoseconds() - g.startTime) / NanosecondsPerReleaseCycle)
	g.release(0, collectionIdx)
}

func (g *GarbageHolder) release(bucket, generationCount int) {
	for generationCount != 0 && bucket < len(g.buckets) {
		remaining := generationCount - g.collected[bucket]
		g.collected[bucket] = generationCount
		if remaining >= BucketSize {
			remaining = BucketSize
		}
		for i := remaining; i < BucketSize; i++ {
			g.buckets[i-remaining] = g.buckets[i] // copy array
			for j := range g.buckets[i] {
				g.buckets[i][j] = nil
			}
		}
		bucket++
		generationCount /= BucketSize
	}
}
